server/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -983,8 +987,25 @@ private Engine.IndexResult applyIndexOperation(
             verifyNotClosed(e);
             return new Engine.IndexResult(e, version, opPrimaryTerm, seqNo, sourceToParse.id());
         }
+        Engine.IndexResult indexResult = index(engine, operation);
+        if (indexResult.isCreated()) {
+            updateFieldHasValueWithFieldsFromParsedDoc(operation.parsedDoc());
+        }
+        return indexResult;
+    }
+
+    private void updateFieldHasValueWithFieldsFromParsedDoc(ParsedDocument document) {
+        document.docs().forEach(doc -> doc.getFields().forEach(field -> setFieldHasValue(field.name())));
```

Comment:
I'll benchmark this today/tomorrow to see how much impact it has end-to-end. It might be more benign if we could at least do indexed for loops here to avoid allocating the lambdas?  That would be my preferred option I think, just to soften the blow in any case :)

Reply:
Sounds good, I removed the lambdas in favour of for loops. We can benchmark it together if you want. :)